support for `AbstractAWSConfig`

ExpandingMan opened the pull request:
This adds support for `AbstractAWSConfig` (see [here](https://github.com/JuliaCloud/AWS.jl/pull/274)) and adds some minimal documentation of `S3Path` to the README.  The changes here are minimal: `AWSConfig` arguments are generalized to `AbstractAWSConfig` and `S3Path` now accepts a single type parameter which is the type of the config.

Note that currently there is a limitation in `FilePathsBase` that prevents types with parameters from being used as paths.  That issue is fixed [here](https://github.com/rofinn/FilePathsBase.jl/pull/111).  `S3Path` will not work properly until that is merged and tagged (it's a very small change so I'm hoping that can happen quite soon).

ExpandingMan commented:
Note that the requisite PR to `FilePathsBase` has already been merged and tagged, so there should be no issue there.

ExpandingMan commented:
AWS has now been tagged, so this should be fully working now.

mattBrzezinski commented on Project.toml:
```diff
@@ -1,6 +1,6 @@
 name = "AWSS3"
 uuid = "1c724243-ef5b-51ab-93f4-b0a88ac62a95"
-version = "0.8.2"
+version = "0.8.3"
```
GH is not allowing me to comment further down in the file.
However, do we not need to bump:

```toml
[compat]
AWS = "1.25"
```

ExpandingMan replied:
done

mattBrzezinski commented on README.md:
````diff
@@ -19,3 +19,44 @@ s3_enable_versioning(aws, "my.bucket")
+[FilePathsBase.jl](https://github.com/rofinn/FilePathsBase.jl), allowing users to use
+Julia's `Base` [file system interface](https://docs.julialang.org/en/v1/base/file/) to
+obtain information from S3 buckets.  See the below example.
+```julia
````
```suggestion

```julia
```

ExpandingMan replied:
done

mattBrzezinski commented on src/AWSS3.jl:
```diff
@@ -7,9 +7,6 @@
 #==============================================================================#
 
 
-__precompile__()
```
Any reason to take this out?

ExpandingMan replied:
It has been unnecessary (precompile true by default, same as `__precompile__()` since Julia version 0.7.  Since the compat bound is 1 anyway, it is not necessary on any version of Julia that this package will work on.  Seemed superfluous.

mattBrzezinski replied:
Ah interesting, didn't know that!

mattBrzezinski commented:
bors try

ExpandingMan commented:
Looks like I broke some of the `S3Path` constructors.  Will fix.

ExpandingMan commented:
We have a bigger problem here, though apparently it's not a new one.

So, if you do `FilePathsBase.register`, this does a `pushfirst!` onto a global variable.  Therefore, any time you do, e.g. `p"/usr/bin/exe"`, the very first thing that will happen is that the program tries to talk to AWS.  This is really bad, for all sorts of reasons, not the least of which is that if you happen to not be on AWS (as is my case here where I'm testing with min.io) this will basically hang for a super long time trying and failing to talk to an EC2 server.

I strongly suggest that this package should *NOT* call `register` by default.  Instead, we should include in the docs that users have the option to do this, but make sure they understand the consequences.

Do you agree with this?  If so, I will make the change in this PR.  Otherwise, I will try to get tests passing, and we can figure it out after this gets merged.

ExpandingMan commented:
Alright, so I think I fixed the error that we originally saw on bors.  I'm a little confused by it, becuase it seems like it must have been an error all along, but regardless, should now be fixed.

Additionally, I've implemented a little hack to make sure that FilePathsBase doesn't inadvertently cause calls to AWS.  The `tryparse` method for `S3Path` now accepts `nothing` as an argument for `config` and this is the new default.  If it gets `nothing`, it will first check if the string is an S3 string before attempting to construct the config.  This prevents code in FilePathsBase from trying to talk to AWS to get to an AWS config unless the user has a `s3://` path string.

If it were up to me, I would still remove the `FilePathsBase.register` call from the `__init__()` of this package.  It just seems a little too weird that `p"s3://bucket/path` can start talking to AWS.  If we removed it users could still call it, but there would be fewer nasty surprises.

In any case, that would be a breaking change, and what I have here now should be non-breaking, and also eliminate any crazy behavior.  Let me know what you think about `register`.  If you want to change it, I can change it now, or we can merge this and tag it and change it afterwards as a breaking change.

I don't know if bors will work for me but I'll give it a shot so we can see results.

ExpandingMan commented:
bors try

ExpandingMan commented:
Ugh, that was because I used `isnothing` which is not available on whatever old version of Julia this was tested on.  I changed it, try again when ready.

mattBrzezinski commented:
bors try

ExpandingMan commented:
Man, doing this without being able to run unit tests locally sure is painful.

Anyway, that was because the `Type` argument to `tryparse` was inappropriately strict.  Should be fixed, try again when ready

mattBrzezinski commented:
bors try

ExpandingMan commented:
Ok, so apparently FilePathsBase utilizes a couple of different constructors called directly with the type (e.g. `typeof(arg)()`).  I have added the appropriate methods.  The constructor methods available may now seem a little messy, but I have tried to make it completely non-breaking while also always handling the conversion for the first argument.

Once more unto the breach?

mattBrzezinski commented:
bors try

mattBrzezinski commented on src/s3path.jl:
```diff
@@ -71,15 +85,18 @@ function S3Path(
 
-function Base.tryparse(::Type{S3Path}, str::AbstractString; config::AWSConfig=global_aws_config())
+# if config=nothing, will not try to talk to AWS until after string is confirmed to be an s3 path
+function Base.tryparse(::Type{<:S3Path}, str::AbstractString; config::Union{Nothing,AbstractAWSConfig}=nothing)
```
Can this not be?

> ; config::AbstractAWSConfig=global_aws_config())

Then remove the `(config === nothing)`

ExpandingMan replied:
No, that's the whole point.

What was happening here was that any time you try to parse a path, that is *ANY* path with `p"/a/b/c"`, because of the way `FilePathsBase.register` works the *very first thing* that would happen (if you don't have AWS environment variables) is that it starts trying to talk to the AWS API to try to get EC2 credentials.

So for example, I was testing this on one of my machines with min.io.  I have no AWS environment variables and am not an AWS's environment.  So, if I do `p"/etc/fstab"` (yes, even a filesystem path) it hangs for a long time and then ultimately errors because it called `global_aws_config()`, did not have local environment variables, and could not talk to the AWS end point where it would get EC2 credentials.  Arguably, this would be bad even on AWS because an unexpectedly expensive network call is taking place even when you are doing something as innocent as declaring a file system path.

The work around you see here partially solves the problem by ensuring that `global_aws_config()` is *NOT* called unless the path it is trying to parse really is an S3 path.

Again, I personally find even this a little dubious, and if it were up to me I'd remove `FilePathsBase.register` from the initialization of AWSS3 (users who wanted it would still have the option of adding it themselves).  However, the changes I have made above are non-breaking, so there's no need to decide whether you agree with me on that right now.

mattBrzezinski replied:
Ah okay, thank you for clarifying! This solution definitely makes sense to go with, and also should lead to some expanse in our unit testing to run a some suite w/o credentials so we can keep track of these things.

ExpandingMan replied:
Great, thanks.  Yeah, some more tests are probably needed.  I'm not going to offer to add any since it's so hard for me to test this package.

Btw, after this is merged, I'm going to publish Minio.jl, which is just a very small package that makes it convenient to declare a config for minio, and also makes it convenient to fetch and run the minio binary as a server.  Part of why I set that up is to make testing easier, and testing this package could be a possible application of this.

mattBrzezinski replied:
Logged an issue for it:

- https://github.com/JuliaCloud/AWSS3.jl/issues/142

I'll get around to it at some point

mattBrzezinski commented on src/s3path.jl:
```diff
@@ -1,14 +1,26 @@
+
+    # this method ensures that there is proper conversion to Tuple{Vararg{String}} for first arg,
+    # which can be a little tempremental if you don't explicitly handle it
+    function S3Path{A}(segments, root::AbstractString, drive::AbstractString, isdirectory::Bool,
```
Is there some more context to this to adding this in? Just not following how the change from `AWSConfig -> AbstractAWSConfig` necessitates this

ExpandingMan replied:
It's rather subtle.  It turns out that `FilePathsBase` calls type constructors directly somewhere.  For example, if you have a function
```julia
f(::Type{T}) where {T} = T()
```
if `T` is `S3Path{AWSConfig}`, this will fail unless the above method is defined, because this is a *different method* from `S3Path()` (because `S3Path{AWSConfig}` and `S3Path` are different types, the latter is an implicit abstract type).

I concede that the constructors I've wound up with here are a little messy, but I have included them because they should
- Ensure there are no breaking changes for the `S3Path` constructor, i.e. the user interface of AWSS3.jl has no breaking changes from this PR.
- Ensure that `FilePathsBase` has the methods it needs for the calls it makes.
- Ensure that the first argument of `S3Path` always gets converted to the correct type, which is a little subtle (due to it being a bit of a weird type) and seemed to be error prone before this PR.

mattBrzezinski commented on src/s3path.jl:
```diff
@@ -1,14 +1,26 @@
+
+    # this method ensures that there is proper conversion to Tuple{Vararg{String}} for first arg,
+    # which can be a little tempremental if you don't explicitly handle it
+    function S3Path{A}(segments, root::AbstractString, drive::AbstractString, isdirectory::Bool,
+                       config::A) where {A<:AbstractAWSConfig}
```
```suggestion
    function S3Path{A}(
        segments,
        root::AbstractString,
        drive::AbstractString,
        isdirectory::Bool,
        config::A
    ) where {A<:AbstractAWSConfig}
```

mattBrzezinski replied:
....GH suggestions just don't seem to be happy. Breaking this into the 2nd multi-line if needed fits better with the style guide

ExpandingMan replied:
That seems rather excessively verbose, but ok, committing.

ExpandingMan commented:
Looks like we are resolved then.  I suggest you run a bors one more time just to sanity check that one syntax change.

After merging, a tag would be hugely appreciated, because like I said, I have a small package I'm waiting on this to register.

Thanks!

mattBrzezinski commented on src/s3path.jl:
```diff
@@ -1,14 +1,31 @@
+        isdirectory::Bool,
+        config::A
+    ) where {A<:AbstractAWSConfig}
+        new{typeof(config)}(tuple((String(s) for s ∈ segments)...), root, drive, isdirectory, config)
```
Just want to cc @rofinn here since he is the mastermind behind this pathing.
Not sure if opinions are had here or not.

ExpandingMan replied:
The issue is that `Tuple{Vararg{String}}` is much stricter than other types (e.g. `Vector`) with conversions, so one has to be careful with e.g. `Tuple{Vararg{SubString}}`.  I considered changing the type of the field, since it's not even a concrete type, but I noticed that this is the same field type as those in `FilePathsBase` so I figured it was best to leave it alone and just make sure it wasn't going to crash.

mattBrzezinski commented on src/s3path.jl:
```diff
@@ -1,14 +1,31 @@
+        isdirectory::Bool,
+        config::A
+    ) where {A<:AbstractAWSConfig}
+        new{typeof(config)}(tuple((String(s) for s ∈ segments)...), root, drive, isdirectory, config)
```
```suggestion
        new{typeof(config)}(tuple((String(s) for s in segments)...), root, drive, isdirectory, config)
```

ExpandingMan replied:
😞

rofinn commented on src/s3path.jl:
```diff
@@ -1,14 +1,31 @@
+    config::A
+
+    # this method ensures that there is proper conversion to Tuple{Vararg{String}} for first arg,
+    # which can be a little tempremental if you don't explicitly handle it
```
```suggestion
    # which can be a little temperamental if you don't explicitly handle it
```

rofinn replied:
Can you clarify what the issue was?

rofinn commented on src/s3path.jl:
```diff
@@ -1,14 +1,31 @@
+        isdirectory::Bool,
+        config::A
+    ) where {A<:AbstractAWSConfig}
+        new{typeof(config)}(tuple((String(s) for s in segments)...), root, drive, isdirectory, config)
```
I don't love that we need to do this iteration, conversion and splatting every time we construct the path type? Can we have an outer constructor that addresses the specific case where this is necessary (e.g., segments aren't `Tuple{Vararg{String}}`?

ExpandingMan replied:
I was in the process of making a simple MWE to show you why it so easily fails, but to my surprise, in my simple example I was not having the issues I had with this that prompted me to make this change.  I'll experiment with this a bit and see if I can come up with a better solution.

rofinn replied:
FWIW, if you find the failure case it might be worth adding a test here and opening an issue against FilePathsBase.jl? Even if we can't fix the problem there generically, we could at least fix the system paths and include it in the test suite.

rofinn reviewed (approved):
Seems mostly fine to me, but I don't like addressing the segments special case in an inner constructor. I think an appropriate dispatch for the outer constructors would suffice.

ExpandingMan commented:
Sorry about that, turns out I had misdiagnosed the problem I was having.  The problem was *not* that Julia was having trouble converting the type of the first argument, the problem was that I was not properly directing the call to the method that converts.

I have removed the inner constructor, and I have added a simple method (without any splatting) to ensure that the constructor still performs proper conversions even if no type parameter is provided.

Note that we seem to still need the `S3Path{T}() where {T}` constructor, as FilePathsBase likes to call constructors like `typeof(arg)()` and I wanted to keep `S3Path()` so as not to have any breaking changes (and might as well have it, right?).

Anyway, should be good now, please trigger bors when ready.

rofinn commented:
> Note that we seem to still need the S3Path{T}() where {T} constructor, as FilePathsBase likes to call constructors like typeof(arg)() and I wanted to keep S3Path() so as not to have any breaking changes (and might as well have it, right?).

Cool, yeah, that makes sense to me :)

ExpandingMan commented:
> FWIW, if you can think of a better way to handle the registration/Path(...)/p_str logic feel free to open an issue. That stuff was very much a means to an end...

I think it's fine like it is, I just think packages (particularly this one) should be more conservative about how they call `FilePathsBase.register`.  Unfortunately using a `@p_str` macro severely restricts how you can pass it context, so I don't really see a better way of doing it than what you did.  It's a nice option to have if you really want it, but I'd think in many situations it would make more sense to just call the constructor.  I suppose you could do something clever in the macro like sniff out the environment, but that seems a little too magical, probably better just to give users more control.

Btw, if you guys are willing to remove `__init__() = FilePathsBase.register(S3Path)` in this PR and just suggest it in documentation, let me know.  In my mind there is no question that it should go, but it would be a very breaking change (anyone who made extensive use of `p"s3://a/b/c` would be in fur a rude update), and I kind of just want to get this merged, so I wasn't going to press the matter.

rofinn commented:
> anyone who made extensive use of p"s3://a/b/c would be in fur a rude update

Yeah, we use it pretty extensively and simply rely config files and environment variables to control the default behaviour.

mattBrzezinski commented:
bors try

mattBrzezinski commented:
Going over the new commits in the morning, and then I'll merge and tag!

mattBrzezinski commented:
bors r+

mattBrzezinski commented:
Ran into an AWS timing issue, I'll retry.

mattBrzezinski commented:
bors r+

ExpandingMan commented:
Thanks!  Just a gentle reminder to please not forget about a tag.  There should be no breaking changes so 0.8.3 should work.

ExpandingMan commented:
Oops, didn't notice that it was already on its way.  Thanks again!

rofinn commented:
Since this PR technically fixes a bug, would it make sense to backport at least the `tryparse` fix to 0.7?

mattBrzezinski commented:
> Since this PR technically fixes a bug, would it make sense to backport at least the `tryparse` fix to 0.7?

Created a PR for this here, https://github.com/JuliaCloud/AWSS3.jl/pull/144/files, not sure the proper way to backport though.

rofinn commented:
I think we typically just make a branch for the release (`0.7.7`)and make the tag off that. You should be able to run the julia registrator off that branch/commit though.

mattBrzezinski commented:
> I think we typically just make a branch for the release (`0.7.7`)and make the tag off that. You should be able to run the julia registrator off that branch/commit though.

Hmmm interesting, we would then need that branch to continuing living forward then outside of `master` I guess?

rofinn commented:
Not necessary. You just need the branch to exist long enough to tag the release. The commit will still exist if it isn't on a branch and folks can always make subsequent branches off of your tagged release.

oxinabox commented:
In retrospect making the config type a type parameter was not a great change.
It was breaking, and didn't give any real gain.

Making it a type parameter, rather than an abstract field avoids dynamic dispatch.
But dynamic dispatch costs 20-200ns.
In contrast, this is doing network IO, so it already takes 20-200us.
